### MATH2504 S2, 2022, BigHW Submission 
### Submission form II of IV 

Student #1 name: Rachel Dudley

Student #1 ID: 46982243

Student #2 name:

Student #2 ID:

GitHub repo: 

---

**Note: It is expected that this notebook be runnable in the sense that the marker can select "Restart & Run All" and all code cells of solutions will execute. This requires not to have name clashes between solutions.**

---

# Solution to [Question 5](https://courses.smp.uq.edu.au/MATH2504/2022/assessment_html/bighw.html#q5)

In [1]:
function is_perfect_number(x::Number)
    factor_ = []
        for i in 2:x
            if x % i == 0
                push!(factor_, x / i)
            end
        end
    sum_ = sum(factor_)
    
    if sum_ == x
            println("$x is a perfect number!")
#     else
#         println("$x is not a perfect number")
    end
end

is_perfect_number (generic function with 1 method)

In [2]:
for A in 2:100000
    is_perfect_number(A)
end

6 is a perfect number!
28 is a perfect number!
496 is a perfect number!
8128 is a perfect number!


---

*Use this area below as an experimental playground for the question. Feel free to leave some of the experimental code here with your submision. However, it will not be marked or reviewed. Make sure in your submission that if there is any experimental code here then running it does not interfere with running the actual solutions.*

---

# Solution to [Question 6](https://courses.smp.uq.edu.au/MATH2504/2022/assessment_html/bighw.html#q6)

In [ ]:
### Answer 6 here (in place of this comment)

---

*Use this area below as an experimental playground for the question. Feel free to leave some of the experimental code here with your submision. However, it will not be marked or reviewed. Make sure in your submission that if there is any experimental code here then running it does not interfere with running the actual solutions.*

---

# Solution to [Question 7](https://courses.smp.uq.edu.au/MATH2504/2022/assessment_html/bighw.html#q7)

In [5]:
using Random
random_license_plate() = String(rand('0':'9',3)) * " " * String(rand('A':'Z', 3))
    
function prepare_data(;N = 10^7)
    Random.seed!(0)
    indexes = rand(1:10^12, N)
    plates = [random_license_plate() for _ in 1:N]
    
    #One specific spot is set to a special value.
    
    special_spot = Int(floor(0.9*N)) #Arbitrary index
    indexes[special_spot], plates[special_spot] = 1234567890, "000 ZZZ"
    
    return indexes, plates
    
end

function prepare_sorted_data(;N = 10^7)
    
    indexes, plates = prepare_data(N=N)
    perm = sortperm(indexes)
    
    return indexes[perm], plates[perm]
end

    indexes, plates = prepare_sorted_data(N=5) #as an example set N=5
@show indexes
@show plates;

indexes = [1234567890, 68545824387, 405699470893, 661612690731, 862140857196]
plates = ["000 ZZZ", "823 LXT", "117 HAO", "571 CJF", "154 AMQ"]


## 7a:

The one line function: (random_license_plate()) generates an array of String values consisting of 3 random numbers from 1 through to 9, and then a space, and then 3 random capital letters from A through to Z. The output could look like "AGF 898".

In this code that generates the ranodm license plates, the 3 numbers, the space, and the 3 letters are joined together through the concatenation of strings via an asterisk *. Some other languages use the + symbol to join strings together, however since addition is always commutative, it is more justified to use an * instead (multiplication is not commutative in all cases e.g. matrix multiplication). For example in line 1 of the code: (rand('0':'9',3)) * " " * String(rand('A':'Z', 3)) ensures the order is always structures such that the 3 numbers come first, followed by a space, followed by the 3 letters.

The use of rand() is to generate values that are random and will then need to be sorted differently each time new license plates are produced. The code also uses Random.seed!(1) so each time rand() is written in the code, a new pseudorandom value will get made and thus the license plates continue to be different each time rand() is called.

The use of a comprehension of the form [for _ in 1:N] tells the code how many times to generate a new licence plate. It runs through each number between 1 and N (this is what for _ in means) and for each one, a new random license plate is produced. For example if the user inputs prepare_data(5), they have set N = 5, and therefore there will be FIVE: (1, 2, 3, 4, 5) loops of random_license_plate().

The tuples being returned in the code are indexes and plates. The return indexes, plates line will return the array indexes, followed by the corresponding array: plates. 

The built-in Julia function: sortperm(), takes a list of values and puts them in lexicographical order, while returning the index of their position in the array. It is used in the prepare_sorted_data() function on the elements in 'indexes'. Since indexes are all numerical values, sortperm(indexes) rearranges all of the indexes so that they are in ascending order and outputs the order they should be in. e.g. sortperm([1, 5, 9, 7, 3]) would return: 1, 5, 2, 4, 3 because the smallest number (1) is in position 1, the next smallest (3) is in position 5, ... and the largest number in the array (9) is in position 3.

## 7b:

The probability of the code returning a repeat index of 1234567890 is N-1/10^12, where N is the number of generated indexes. This is because there are N random indexes generated and they each have the possibility to be between 1 and 10^12. Since the special index 1234567890 will always be present as one of the N generated indexes, there are N-1 possible spots which could be randomly generated as 1234567890. E.g if there are 3 indexes generated, 1 of them will definitely be the special index, and then there is a 2/10^12 chance that one of the other indexes will match that special index. If N could be (10^12 +1) then the probability of getting a repeat of 1234567890 would be 10^12/10^12 = 1 so it is pretty much certain that there would be a repeat. In this case however, 10^7-1/10^12 = 0.00001 is the probability which is very small

The package has had 306 commits

One issue in the repo is:  Improper axis limits in histograms #253 which is about where the x axis should have its limits set to. The user who mentions the issue suggests that the axis labels should always be set to 0 on the LHS, and be bounded on the right by a whole power of 10 seconds (as opposed to being cut off at the x value where the last data value).

the difference between @benchmark and @btime is: @benchmark defines the benchmark, auto-tunes its configuration parameters and runs the benchmark. It also gives a more detailed summary of time taken compared to @btime for each of the samples run, including minimum, maximum, median, mean and sd. On the other hand, @btime prints only the minimum time and the memory allocation. Then it returns the output of what is being @btime'd.

In [9]:
using BenchmarkTools

function find_special_plate(indexes, plates; special_index = 1234567890)
    for (i, index) in enumerate(indexes)
        if index == special_index
            return plates[i]
        end
    end
    
throw(Error("Index not found"))
    
end

indexes, plates = prepare_sorted_data()
#benchmark find_special_plate(indexes, plates)
@time find_special_plate(indexes, plates)

  0.023628 seconds (21.78 k allocations: 1.143 MiB, 99.80% compilation time)


"000 ZZZ"

## 7c:

The function find_special_plate returned 0.013943 seconds (21.78 k allocations: 1.143 MiB, 99.78% compilation time) - when using the @time benchmark macro, whereas find_special_plate_binary returned 0.000229 seconds (18 allocations: 592 bytes) which is much faster. the binary search has shown in this example to be 2 orders of magnitude faster then the inital search method. The binary search method also has significantly less memory allocations which is optimal.

In [7]:
using BenchmarkTools

function find_special_plate_binary(indexes, special_index, min, max)
    
    if min > max
        
        throw("error")
        
    else
        middle_value = trunc(Int, floor((min+max)/2))
        
        if special_index == indexes[middle_value]
        
        print("the value at indexes[$(middle_value)] is: $(indexes[middle_value])")
        return middle_value 
            
    elseif special_index > indexes[middle_value]
        
        return find_special_plate_binary(indexes, special_index, middle_value+1, max)
            
    else
        
        return find_special_plate_binary(indexes, special_index, min, middle_value-1)
        
    end
    end
    
end

#indexes, plates = prepare_sorted_data()
#find_special_plate_binary(indexes, 1234567890, 1, 10^7)

find_special_plate_binary (generic function with 1 method)

In [10]:
@time find_special_plate_binary(indexes, 1234567890, 2, 10^7)

the value at indexes[12505] is: 1234567890  0.000229 seconds (18 allocations: 592 bytes)


12505

## 7d:

The time complexity for binary search is O(log n). This means that as the input size of the function grows, the number of operations performed grows as a significantly smaller rate. -> as n in log(n) grows by x10, log(n) only grows by +1. Another feature making O(logn) much faster, is that it does not search through every value in the array, it only searches for a location in the array and then shortens its search scope depending on the value of the element in that location. 

---

*Use this area below as an experimental playground for the question. Feel free to leave some of the experimental code here with your submision. However, it will not be marked or reviewed. Make sure in your submission that if there is any experimental code here then running it does not interfere with running the actual solutions.*

---

# Solution to [Question 8](https://courses.smp.uq.edu.au/MATH2504/2022/assessment_html/bighw.html#q8)

## 8a:

In [12]:
sort(names, by= (x) ->length(split(x," ")[end]))

LoadError: MethodError: no method matching sort(::typeof(names); by=var"#10#11"())
Closest candidates are:
  sort(::AbstractUnitRange) at C:\Users\kevin\.julia\juliaup\julia-1.7.3+0~x64\share\julia\base\range.jl:1304 got unsupported keyword argument "by"
  sort(::AbstractRange) at C:\Users\kevin\.julia\juliaup\julia-1.7.3+0~x64\share\julia\base\range.jl:1307 got unsupported keyword argument "by"
  sort(::SparseArrays.SparseVector{Tv, Ti}; kws...) where {Tv, Ti} at C:\Users\kevin\.julia\juliaup\julia-1.7.3+0~x64\share\julia\stdlib\v1.7\SparseArrays\src\sparsevector.jl:1968
  ...

## 8b:

In [3]:
#sort(names, by= (x) ->length(split(x," ")[end]))
#last_names = Array{String}(undef, 0)
names = [   "Amy Chan",
            "Maithili Mehta",
            "Anna Foeglein",
            "Andy Ferris",
            "Thomas Graham",
            "Elaine Schenk",
            "Jesse Woods",
            "Tina Moghaddam",
            "Paul Bellette",
            "Paul Vrbik",
            "Tom Cranitch",
            "Yoni Nazarathy",
            "Sam Hambleton",
            "Alistair Falconer",
            "Emma Comino",
            "Ivana Carrizo-Molina"]

#sort(names, by= (x) ->length(split(x," ")[end]))
#last_names = Array{String}(undef, 0)
function my_sort_bubble!(a; by= (a)-> length((a)))
    n = length(a)
    for i in 1:n-1
        for j in 1:n-i
            if  by(a[j]) > by(a[j+1])    
                a[j], a[j + 1] = a[j + 1], a[j]
            end
        end
    end
end 
my_sort_bubble!(names,by= (a)->length(split(a," ")[end]))
println("The name order by the length of last names:", names)

LoadError: cannot assign a value to variable Base.names from module Main

## 8c:

In [ ]:
#we sorted it once, how do we sort it twice or more? 
#I am sure that part_sort_quick is working
function part_sort_quick(a, l, r) #a is the array, l is the indexes left to the pivot and r is the pivot right to the pivot
    pivot = a[r] #define the pivot
    i = l - 1  #initalize the starting position of i
        for j ∈ l:r-1 #for k from the first element to the last element before the pivot
            if a[j]<pivot
                i += 1 #increment i by 1
                a[i],a[j] = a[j],a[i] #swap the element at i and j
            end
        end
        a[i+1] ,a[r] = a[r], a[i+1]
    #swap the pivot with i+1 element so that all values on the left of the pivot is smaller than pivot
    return i + 1 #return index of the pivot
end
#quick sort should be a recursive function
function my_quick_sort(a,l,r)
    if l >= r #if the l is bigger or equal to the right there is nothing more to sort
        return
    else l < r
            p = part_sort_quick(a,l,r)
            @show part_sort_quick(a,l,r)
            my_quick_sort(a, l, p-1)
            my_quick_sort(a, p+1, r) 
        return a
    end 
end
#

## 8d:

In [ ]:
using BenchmarkTools
data = rand(Int, 10^6)
data_1 = rand(Int, 10^5)
@btime sort(data)
@btime my_quick_sort(data_1,1,length(data_1))

---

*Use this area below as an experimental playground for the question. Feel free to leave some of the experimental code here with your submision. However, it will not be marked or reviewed. Make sure in your submission that if there is any experimental code here then running it does not interfere with running the actual solutions.*

---

10-element Vector{Float64}:
 0.3715278714713923
 0.5827838677957287
 0.5001923764946207
 0.6194998257401508
 0.024160832819473677
 0.9234086929567886
 0.04930050236040784
 0.26820010716880116
 0.644552935257493
 0.6047031796255372

# Solution to [Question 9](https://courses.smp.uq.edu.au/MATH2504/2022/assessment_html/bighw.html#q9)

In [1]:
using Random
using LinearAlgebra
Random.seed!(0)
@show A = round.(10*rand(2,3))
@show B = round.(10*rand(3,4))
A*B

A = round.(10 * rand(2, 3)) = [4.0 9.0 7.0; 1.0 1.0 1.0]
B = round.(10 * rand(3, 4)) = [7.0 7.0 1.0 9.0; 10.0 4.0 3.0 9.0; 4.0 5.0 7.0 8.0]


2×4 Matrix{Float64}:
 146.0  99.0  80.0  173.0
  21.0  16.0  11.0   26.0

## 9a:

In [2]:
using BenchmarkTools
function my_mult_by_cols(A, B)
    
    nA,mA = size(A)
    nB,mB = size(B)
    @assert mA == nB
    n, m, p, q = nA, mA, mB, nB
    C = Array{Float64}(undef,n,p)
    
    for c_columns in 1:p
            
        C[:, c_columns] = sum([A[:, a_columns] * B[a_columns, c_columns] for a_columns in 1:m])
        
    end
    
    return C
    
end

my_mult_by_cols(A, B)

2×4 Matrix{Float64}:
 146.0  99.0  80.0  173.0
  21.0  16.0  11.0   26.0

## 9b:

In [3]:
function my_mult_by_rows(A, B)

    nA,mA = size(A)
    nB,mB = size(B)
    @assert mA == nB
    n, m, p, q = nA, mA, mB, nB
    C = Array{Float64}(undef,n,p)
    
    for a_rows in 1:n
            
        C[a_rows, :] = sum([A[a_rows, b_cols] * B[b_cols, :] for b_cols in 1:q])
        
    end
    
    return C
    
end

my_mult_by_rows(A, B)

2×4 Matrix{Float64}:
 146.0  99.0  80.0  173.0
  21.0  16.0  11.0   26.0

## 9c:

In [3]:
function my_mult_sum_outer_products(A, B)
 
nA,mA = size(A)
nB,mB = size(B)
@assert mA == nB
n, m, p, q = nA, mA, mB, nB
outer_products = Array{Float64}(undef,n,p)    
    
    for i in 1:m
        
        outer_products += A[:, i] * B[i, :]'
    
    end
    
#     a1b1 = A[:, 1] * B[1, :]'
#     a2b2 = A[:, 2] * B[2, :]'
#     a3b3 = A[:, 3] * B[3, :]'
    
    #outer_product_sum = a1b1 + a2b2 + a3b3
        
    return outer_products
        
end

@time my_mult_sum_outer_products(A, B)

  0.055752 seconds (32.18 k allocations: 1.747 MiB, 99.95% compilation time)


2×4 Matrix{Float64}:
 146.0  99.0  80.0  173.0
  21.0  16.0  11.0   26.0

## 9d:


In terms of complexity of the numebr of operations, single for loops are far more efficient then nested for loops. This is because using a nested for loop is going through the operations 2 full times (for each of the for loops) which is doubling the number of computations the function needs to do. It is always best to avoid nested loops where possible if they can be replaced with a single loop as this will end up computing far less operations which is more efficient 

---

*Use this area below as an experimental playground for the question. Feel free to leave some of the experimental code here with your submision. However, it will not be marked or reviewed. Make sure in your submission that if there is any experimental code here then running it does not interfere with running the actual solutions.*

---

In [15]:
function my_mult_inner_products(A,B)
nA,mA = size(A)

nB,mB = size(B)
@assert mA == nB
n, m, p = nA, mA, mB
C = Array{Float64}(undef,n,p)
for i in 1:n
for j in 1:p
C[i,j] = A[i,:]' * B[:,j] #compute inner product of i'th row of A a
end
end
return C
end
@time my_mult_inner_products(A,B)

  0.156127 seconds (81.01 k allocations: 4.456 MiB, 98.95% compilation time)


2×4 Matrix{Float64}:
 146.0  99.0  80.0  173.0
  21.0  16.0  11.0   26.0